src/thermo/PengRobinsonMFTP.cpp

```diff
@@ -0,0 +1,1321 @@
+    };
+
+    boost::uintmax_t maxiter = 100;
+    std::pair<double, double> vv = bmt::toms748_solve(
+        resid, Vroot[1], Vroot[2], bmt::eps_tolerance<double>(48), maxiter);
+
+    doublereal mmw = meanMolecularWeight();
+    return mmw / (0.5 * (vv.first + vv.second));
+}
+
+doublereal PengRobinsonMFTP::pressureCalc(doublereal TKelvin, doublereal molarVol) const
+{
+		doublereal den = molarVol * molarVol + 2 * molarVol * m_b_current - m_b_current * m_b_current;
+    double pres = GasConstant * TKelvin / (molarVol - m_b_current) - m_aAlpha_current / den;
+    return pres;
+}
+
+doublereal PengRobinsonMFTP::dpdVCalc(doublereal TKelvin, doublereal molarVol, doublereal& presCalc) const
+{
+		doublereal den = molarVol * molarVol + 2 * molarVol * m_b_current - m_b_current * m_b_current;
+		presCalc = GasConstant * TKelvin / (molarVol - m_b_current) - m_aAlpha_current/ den;
+
+    doublereal vpb = molarVol + m_b_current;
+    doublereal vmb = molarVol - m_b_current;
+		doublereal dpdv = -GasConstant * TKelvin / (vmb * vmb) + 2 *m_aAlpha_current * vpb / (den*den);
+    return dpdv;
+}
+
+void PengRobinsonMFTP::pressureDerivatives() const
+{
+    doublereal TKelvin = temperature();
+    doublereal mv = molarVolume();
+    doublereal pres;
+
+    dpdV_ = dpdVCalc(TKelvin, mv, pres);
+    doublereal vmb = mv - m_b_current;
+		doublereal den = mv * mv + 2 * mv * m_b_current - m_b_current * m_b_current;
+		dpdT_ = (GasConstant / vmb - daAlpha_dT() / den);
+}
+
+void PengRobinsonMFTP::updateMixingExpressions()
+{
+    updateAB();
+}
+
+void PengRobinsonMFTP::updateAB()
+{
+    double temp = temperature();
+		//Update aAlpha_i
+		double sqt_alpha;
+		double criTemp = critTemperature();
+		double sqt_T_reduced = sqrt(temp / criTemp);
+
+		// Update indiviual alpha
+		for (size_t j = 0; j < m_kk; j++) {
+			sqt_alpha = 1 + kappa_vec_[j] * (1 - sqt_T_reduced);
+			alpha_vec_Curr_[j] = sqt_alpha*sqt_alpha;
+		}
+
+		//Update aAlpha_i,j
+		for (size_t i = 0; i < m_kk; i++) {
+				for (size_t j = 0; j < m_kk; j++) {
+						size_t counter = i * m_kk + j;
+						a_vec_Curr_[counter] = a_coeff_vec(0, counter);
+						aAlpha_vec_Curr_[counter] = sqrt(alpha_vec_Curr_[i] * alpha_vec_Curr_[j]) * a_coeff_vec(0, counter);
+				}
+		}
+
+    m_b_current = 0.0;
+    m_a_current = 0.0;
+		m_aAlpha_current = 0.0;
+
+    for (size_t i = 0; i < m_kk; i++) {
+				m_b_current += moleFractions_[i] * b_vec_Curr_[i];
+        for (size_t j = 0; j < m_kk; j++) {
+            m_a_current += a_vec_Curr_[i * m_kk + j] * moleFractions_[i] * moleFractions_[j];
+						m_aAlpha_current += aAlpha_vec_Curr_[i * m_kk + j] * moleFractions_[i] * moleFractions_[j];
+        }
+    }
+}
+
+void PengRobinsonMFTP::calculateAB(doublereal temp, doublereal& aCalc, doublereal& bCalc, doublereal& aAlphaCalc) const
+{
+		bCalc = 0.0;
+		aCalc = 0.0;
+		aAlphaCalc = 0.0;
+		for (size_t i = 0; i < m_kk; i++) {
+				bCalc += moleFractions_[i] * b_vec_Curr_[i];
+				for (size_t j = 0; j < m_kk; j++) {
```

Comment:
You are right. The only difference between the two is an additional calculation of alpha coefficients in PengRobinson class. The easiest way to reduce the duplication will be to define this function in MixtureFugacityMFTP class.  But I fear that it may introduce some unknown implications. Also this function may not work if some other equation of state is implemented in future.

Reply:
This is fixed in https://github.com/Cantera/cantera/pull/641/commits/8f8af55bd02f8d111b8873f4a585e85819b5f27c - I replace those lines with a call to `calculateAB()` (which has replaced `updateAB()`).